librad/src/net/peer.rs

```diff
@@ -535,8 +549,13 @@ where
 
                         match res {
                             Ok(()) => {
+                                // If we cloned from the provider we need to ask for their
+                                // commit.
+                                let mut has = has.clone();
+                                let peer_id = has.origin.unwrap_or_else(|| provider);
```

Comment:
This was already done on line 526 above -- how about constructing the `urn` there and just call `this.git_has` directly?